File: app/src/main/java/com/aricneto/twistytimer/fragment/dialog/TimeDialog.java
package com.aricneto.twistytimer.fragment.dialog;

import android.annotation.SuppressLint;
import android.content.Context;
import android.content.Intent;
import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.graphics.drawable.Drawable;
import android.os.AsyncTask;
import android.os.Bundle;
import android.preference.PreferenceManager;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.view.menu.MenuBuilder;
import androidx.appcompat.view.menu.MenuPopupHelper;
import androidx.fragment.app.DialogFragment;
import androidx.appcompat.widget.PopupMenu;
import android.text.Html;
import android.text.InputType;
import android.view.LayoutInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.inputmethod.EditorInfo;
import android.widget.EditText;
import android.widget.Toast;

import com.aricneto.twistify.R;
import com.aricneto.twistify.databinding.DialogTimeDetailsBinding;
import com.aricneto.twistytimer.TwistyTimer;
import com.aricneto.twistytimer.database.DatabaseHandler;
import com.aricneto.twistytimer.items.Solve;
import com.aricneto.twistytimer.listener.DialogListener;
import com.aricneto.twistytimer.utils.AnimUtils;
import com.aricneto.twistytimer.utils.PuzzleUtils;
import com.aricneto.twistytimer.utils.ScrambleGenerator;
import com.aricneto.twistytimer.utils.TTIntent;
import com.aricneto.twistytimer.utils.ThemeUtils;
import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.textfield.TextInputEditText;

import org.joda.time.DateTime;

/**
 * Shows the timeList dialog
 */
public class TimeDialog extends DialogFragment {

    private DialogTimeDetailsBinding binding;

    private long            mId;
    private Solve           solve;
    private DialogListener  dialogListener;

    @SuppressLint("RestrictedApi")
    private View.OnClickListener clickListener = new View.OnClickListener() {
        @Override
        public void onClick(View view) {
            final DatabaseHandler dbHandler = TwistyTimer.getDBHandler();

            switch (view.getId()) {
                case R.id.overflowButton:
                    PopupMenu popupMenu = new PopupMenu(getActivity(), binding.overflowButton);
                    if (solve.isHistory())
                        popupMenu.getMenuInflater().inflate(R.menu.menu_list_detail_history, popupMenu.getMenu());
                    else
                        popupMenu.getMenuInflater().inflate(R.menu.menu_list_detail, popupMenu.getMenu());

                    MenuPopupHelper popupHelper = new MenuPopupHelper(mContext, (MenuBuilder) popupMenu.getMenu(), binding.overflowButton);
                    popupHelper.setForceShowIcon(true);

                    popupMenu.setOnMenuItemClickListener(new PopupMenu.OnMenuItemClickListener() {
                        @Override
                        public boolean onMenuItemClick(MenuItem item) {
                            switch (item.getItemId()) {
                                case R.id.share:
                                    Intent shareIntent = new Intent();
                                    shareIntent.setAction(Intent.ACTION_SEND);
                                    shareIntent.putExtra(Intent.EXTRA_TEXT, PuzzleUtils.convertTimeToString(solve.getTime(), PuzzleUtils.FORMAT_DEFAULT) + "s.\n" + solve.getComment() + "\n" + solve.getScramble());
                                    shareIntent.setType("text/plain");
                                    getContext().startActivity(shareIntent);
                                    break;
                                case R.id.remove:
                                    dbHandler.deleteSolveByID(mId);
                                    updateList();
                                    break;
                                case R.id.history_to:
                                    solve.setHistory(true);
                                    Toast.makeText(getContext(), getString(R.string.sent_to_history), Toast.LENGTH_SHORT).show();
                                    dbHandler.updateSolve(solve);
                                    updateList();
                                    dismiss();
                                    break;
                                case R.id.history_from:
                                    solve.setHistory(false);
                                    Toast.makeText(getContext(), getString(R.string.sent_to_session), Toast.LENGTH_SHORT).show();
                                    dbHandler.updateSolve(solve);
                                    updateList();
                                    dismiss();
                                    break;
                            }
                            return true;
                        }
                    });
                    popupHelper.show();
                    break;
                case R.id.editButton:
                    new MaterialAlertDialogBuilder(mContext)
                            .setTitle(R.string.select_penalty)
                            .setSingleChoiceItems(R.array.array_penalties, solve.getPenalty(), (dialog, which) -> {
                                switch (which) {
                                    case 0: // No penalty
                                        solve = PuzzleUtils.applyPenalty(solve, PuzzleUtils.NO_PENALTY);
                                        break;
                                    case 1: // +2
                                        solve = PuzzleUtils.applyPenalty(solve, PuzzleUtils.PENALTY_PLUSTWO);
                                        break;
                                    case 2: // DNF
                                        solve = PuzzleUtils.applyPenalty(solve, PuzzleUtils.PENALTY_DNF);
                                        break;
                                }
                                dbHandler.updateSolve(solve);
                                updateList();
                                dialog.dismiss();
                            })
                            .setNegativeButton(R.string.action_cancel, null)
                            .show();
                    break;
                case R.id.commentButton:
                    View commentView = LayoutInflater.from(mContext).inflate(R.layout.dialog_input, null);
                    TextInputEditText commentEditText = commentView.findViewById(R.id.edit_text);
                    commentEditText.setText(solve.getComment());

                    new MaterialAlertDialogBuilder(mContext)
                            .setTitle(R.string.edit_comment)
                            .setView(commentView)
                            .setPositiveButton(R.string.action_done, (dialog1, which) -> {
                                solve.setComment(commentEditText.getText().toString());
                                dbHandler.updateSolve(solve);
                                Toast.makeText(getContext(), getString(R.string.added_comment), Toast.LENGTH_SHORT).show();
                                updateList();
                            })
                            .setNegativeButton(R.string.action_cancel, null)
                            .show();
                    break;
                case R.id.scrambleText:
                    AnimUtils.toggleContentVisibility(binding.scrambleImage);
                    break;
            }
        }
    };
    private Context mContext;

    public static TimeDialog newInstance(long id) {
        TimeDialog timeDialog = new TimeDialog();
        Bundle args = new Bundle();
        args.putLong("id", id);
        timeDialog.setArguments(args);
        return timeDialog;
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setStyle(DialogFragment.STYLE_NO_TITLE, 0);
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        binding = DialogTimeDetailsBinding.inflate(inflater, container, false);
        //this.setEnterTransition(R.anim.activity_slide_in);
        mContext = getContext();

        mId = getArguments().getLong("id");

        //Log.d("TIME DIALOG", "mId: " + mId + "\nexists: " + handler.idExists(mId));

        getDialog().getWindow().setBackgroundDrawable(new ColorDrawable(Color.TRANSPARENT));
        //getDialog().getWindow().setWindowAnimations(R.style.DialogAnimationScale);

        final Solve matchedSolve = TwistyTimer.getDBHandler().getSolve(mId);

        if (matchedSolve != null) {
            solve = matchedSolve;

            binding.timeText.setText(Html.fromHtml(PuzzleUtils.convertTimeToString(solve.getTime(), PuzzleUtils.FORMAT_SMALL_MILLI)));
            binding.dateText.setText(new DateTime(solve.getDate()).toString("d MMM y'\n'H':'mm"));

            binding.scrambleText.setText(solve.getScramble());

            if (solve.getPenalty() == PuzzleUtils.PENALTY_DNF)
                binding.puzzlePenaltyText.setText("DNF");
            else if (solve.getPenalty() == PuzzleUtils.PENALTY_PLUSTWO)
                binding.puzzlePenaltyText.setText("+2");
            else
                binding.puzzlePenaltyText.setVisibility(View.GONE);

            if (solve.getComment() != null) {
                if (! solve.getComment().equals("")) {
                    binding.commentText.setText(solve.getComment());
                    binding.commentText.setVisibility(View.VISIBLE);
                }
            }

            if (solve.getScramble() != null) {
                if (solve.getScramble().equals(""))
                    binding.scrambleText.setVisibility(View.GONE);
            }

            binding.scrambleText.setOnClickListener(clickListener);
            binding.overflowButton.setOnClickListener(clickListener);
            binding.editButton.setOnClickListener(clickListener);
            binding.commentButton.setOnClickListener(clickListener);

        }

        return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);
        new GenerateScrambleImage().execute();
    }

    public void setDialogListener(DialogListener listener) {
        dialogListener = listener;
    }

    private void updateList() {
        if (dialogListener != null) {
            dialogListener.onUpdateDialog();
        } else {
            TTIntent.broadcast(TTIntent.CATEGORY_TIME_DATA_CHANGES, TTIntent.ACTION_TIMES_MODIFIED);
        }
        dismiss();
    }

    @Override
    public void onDestroyView() {
        binding = null;
        if (dialogListener != null)
            dialogListener.onDismissDialog();
        super.onDestroyView();
    }

    private class GenerateScrambleImage extends AsyncTask<Void, Void, Drawable> {

        @Override
        protected Drawable doInBackground(Void... voids) {
            ScrambleGenerator generator = new ScrambleGenerator(solve.getPuzzle());
            return generator.generateImageFromScramble(
                    PreferenceManager.getDefaultSharedPreferences(TwistyTimer.getAppContext()),
                    solve.getScramble());
        }

        @Override
        protected void onPostExecute(Drawable drawable) {
            super.onPostExecute(drawable);
            if (binding != null && binding.scrambleImage != null)
                binding.scrambleImage.setImageDrawable(drawable);
        }
    }

}

File: app/src/main/java/com/aricneto/twistytimer/fragment/dialog/TimeDialog.kt
package com.aricneto.twistytimer.fragment.dialog

import android.content.Context
import android.content.DialogInterface
import android.content.Intent
import android.graphics.Color
import android.os.Bundle
import android.text.Html
import android.util.Log
import android.util.TypedValue
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.appcompat.widget.PopupMenu
import androidx.core.graphics.drawable.toDrawable
import androidx.core.text.HtmlCompat.FROM_HTML_MODE_LEGACY
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.lifecycleScope
import androidx.preference.PreferenceManager
import com.aricneto.twistify.R
import com.aricneto.twistify.databinding.DialogTimeDetailsBinding
import com.aricneto.twistytimer.TwistyTimer
import com.aricneto.twistytimer.items.Solve
import com.aricneto.twistytimer.listener.DialogListener
import com.aricneto.twistytimer.utils.AnimUtils.toggleContentVisibility
import com.aricneto.twistytimer.utils.PuzzleUtils
import com.aricneto.twistytimer.utils.PuzzleUtils.convertTimeToString
import com.aricneto.twistytimer.utils.ScrambleGenerator
import com.aricneto.twistytimer.utils.TTIntent
import com.aricneto.twistytimer.utils.TTIntent.broadcast
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.google.android.material.textfield.TextInputEditText
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext
import kotlinx.datetime.Instant
import kotlinx.datetime.TimeZone
import kotlinx.datetime.toJavaLocalDateTime
import kotlinx.datetime.toLocalDateTime
import java.time.format.DateTimeFormatter
import java.util.Locale

/**
 * Shows the timeList dialog
 */
class TimeDialog : DialogFragment() {
    private var binding: DialogTimeDetailsBinding? = null

    private var mId: Long = 0
    private var solve: Solve? = null
    private var dialogListener: DialogListener? = null

    private val clickListener = View.OnClickListener { view ->
        val solveRepository = TwistyTimer.getSolveRepository()

        when (view.id) {
            R.id.overflowButton -> {
                val popupMenu = PopupMenu(requireActivity(), binding!!.overflowButton)
                if (solve?.history ?: false) popupMenu.menuInflater
                    .inflate(R.menu.menu_list_detail_history, popupMenu.menu)
                else popupMenu.menuInflater
                    .inflate(R.menu.menu_list_detail, popupMenu.menu)

                try {
                    val fieldPopup = PopupMenu::class.java.getDeclaredField("mPopup")
                    fieldPopup.isAccessible = true
                    val menuPopupHelper = fieldPopup.get(popupMenu)
                    val setForceIcons = menuPopupHelper.javaClass
                        .getDeclaredMethod("setForceShowIcon", Boolean::class.javaPrimitiveType)
                    setForceIcons.invoke(menuPopupHelper, true)
                } catch (e: Exception) {
                    Log.e("TimeDialog", "Error forcing icons in PopupMenu: $e")
                }

                popupMenu.setOnMenuItemClickListener { item ->
                    when (item.itemId) {
                        R.id.share -> {
                            val shareIntent = Intent()
                            shareIntent.action = Intent.ACTION_SEND
                            shareIntent.putExtra(
                                Intent.EXTRA_TEXT,
                                convertTimeToString(
                                    solve!!.time.toLong(),
                                    PuzzleUtils.FORMAT_DEFAULT
                                ) + "s.\n" + solve!!.comment + "\n" + solve!!.scramble
                            )
                            shareIntent.type = "text/plain"
                            requireContext().startActivity(shareIntent)
                        }

                        R.id.remove -> {
                            lifecycleScope.launch {
                                solveRepository.deleteSolve(mId)
                                updateList()
                            }
                        }

                        R.id.history_to -> {
                            solve!!.history = true
                            Toast.makeText(
                                context,
                                getString(R.string.sent_to_history),
                                Toast.LENGTH_SHORT
                            ).show()
                            lifecycleScope.launch {
                                solveRepository.updateSolve(
                                    id = solve!!.id,
                                    time = solve!!.time.toLong(),
                                    date = solve!!.date,
                                    scramble = solve!!.scramble,
                                    penalty = solve!!.penalty.toLong(),
                                    comment = solve!!.comment,
                                    history = solve!!.history,
                                    mode = solve!!.mode
                                )
                                updateList()
                            }
                        }

                        R.id.history_from -> {
                            solve!!.history = false
                            Toast.makeText(
                                context,
                                getString(R.string.sent_to_session),
                                Toast.LENGTH_SHORT
                            ).show()
                            lifecycleScope.launch {
                                solveRepository.updateSolve(
                                    id = solve!!.id,
                                    time = solve!!.time.toLong(),
                                    date = solve!!.date,
                                    scramble = solve!!.scramble,
                                    penalty = solve!!.penalty.toLong(),
                                    comment = solve!!.comment,
                                    history = solve!!.history,
                                    mode = solve!!.mode
                                )
                                updateList()
                            }
                        }
                    }
                    true
                }
                popupMenu.show()
            }

            R.id.editButton ->
                MaterialAlertDialogBuilder(requireActivity())
                    .setTitle(R.string.select_penalty)
                    .setSingleChoiceItems(
                        R.array.array_penalties,
                        solve!!.penalty
                    ) { dialog: DialogInterface?, which: Int ->
                        when (which) {
                            0 -> solve =
                                PuzzleUtils.applyPenalty(solve!!, PuzzleUtils.NO_PENALTY)

                            1 -> solve =
                                PuzzleUtils.applyPenalty(solve!!, PuzzleUtils.PENALTY_PLUSTWO)

                            2 -> solve =
                                PuzzleUtils.applyPenalty(solve!!, PuzzleUtils.PENALTY_DNF)
                        }
                        lifecycleScope.launch {
                            solveRepository.updateSolve(
                                id = solve!!.id,
                                time = solve!!.time.toLong(),
                                date = solve!!.date,
                                scramble = solve!!.scramble,
                                penalty = solve!!.penalty.toLong(),
                                comment = solve!!.comment,
                                history = solve!!.history,
                                mode = solve!!.mode
                            )
                            updateList()
                            dialog!!.dismiss()
                        }
                    }
                    .setNegativeButton(R.string.action_cancel, null)
                    .show()

            R.id.commentButton -> {
                val commentView =
                    layoutInflater.inflate(
                        R.layout.dialog_input,
                        null
                    )
                val commentEditText = commentView.findViewById<TextInputEditText>(R.id.edit_text)
                commentEditText.setText(solve!!.comment)

                MaterialAlertDialogBuilder(requireActivity())
                    .setTitle(R.string.edit_comment)
                    .setView(commentView)
                    .setPositiveButton(
                        R.string.action_done
                    ) { _: DialogInterface?, _: Int ->
                        solve!!.comment = commentEditText.text.toString()
                        lifecycleScope.launch {
                            solveRepository.updateSolve(
                                id = solve!!.id,
                                time = solve!!.time.toLong(),
                                date = solve!!.date,
                                scramble = solve!!.scramble,
                                penalty = solve!!.penalty.toLong(),
                                comment = solve!!.comment,
                                history = solve!!.history,
                                mode = solve!!.mode
                            )
                            Toast.makeText(
                                context,
                                getString(R.string.added_comment),
                                Toast.LENGTH_SHORT
                            ).show()
                            updateList()
                        }
                    }
                    .setNegativeButton(R.string.action_cancel, null)
                    .show()
            }

            R.id.scrambleText -> toggleContentVisibility(binding!!.scrambleImage)
        }
    }
    private var mContext: Context? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setStyle(STYLE_NO_TITLE, 0)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        binding = DialogTimeDetailsBinding.inflate(inflater, container, false)
        //this.setEnterTransition(R.anim.activity_slide_in);
        mContext = context

        mId = requireArguments().getLong("id")

        dialog!!.window!!.setBackgroundDrawable(Color.TRANSPARENT.toDrawable())

        lifecycleScope.launch {
            val matchedSolve = TwistyTimer.getSolveRepository().getSolve(mId)

            if (matchedSolve != null) {
                solve = matchedSolve

                binding!!.timeText.text = Html.fromHtml(
                    convertTimeToString(
                        solve!!.time.toLong(),
                        PuzzleUtils.FORMAT_SMALL_MILLI
                    ), FROM_HTML_MODE_LEGACY
                )
                val dateTime = Instant.fromEpochMilliseconds(solve!!.date).toLocalDateTime(TimeZone.currentSystemDefault())
                val formatter = DateTimeFormatter.ofPattern("d MMM y'\n'H':'mm", Locale.getDefault())
                binding!!.dateText.text = dateTime.toJavaLocalDateTime().format(formatter)

                binding!!.scrambleText.text = solve!!.scramble

                when (solve!!.penalty) {
                    PuzzleUtils.PENALTY_DNF -> binding!!.puzzlePenaltyText.text =
                        getString(R.string.do_not_finished)

                    PuzzleUtils.PENALTY_PLUSTWO -> binding!!.puzzlePenaltyText.text =
                        "+2"

                    else -> binding!!.puzzlePenaltyText.visibility = View.GONE
                }

                if (solve!!.comment != "") {
                    binding!!.commentText.text = solve!!.comment
                    binding!!.commentText.visibility = View.VISIBLE
                }

                if (solve!!.scramble == "") binding!!.scrambleText.visibility = View.GONE

                binding!!.scrambleText.setOnClickListener(clickListener)
                binding!!.overflowButton.setOnClickListener(clickListener)
                binding!!.editButton.setOnClickListener(clickListener)
                binding!!.commentButton.setOnClickListener(clickListener)

                // Generate scramble image
                val generator = ScrambleGenerator(solve!!.puzzle)
                val drawable = withContext(Dispatchers.IO) {
                    generator.generateImageFromScramble(
                        PreferenceManager.getDefaultSharedPreferences(requireContext()),
                        solve!!.scramble
                    )
                }
                binding!!.scrambleImage.setImageDrawable(drawable)
            }
        }

        return binding!!.getRoot()
    }

    fun setDialogListener(listener: DialogListener?) {
        dialogListener = listener
    }

    private fun updateList() {
        if (dialogListener != null) {
            dialogListener!!.onUpdateDialog()
        } else {
            broadcast(TTIntent.CATEGORY_TIME_DATA_CHANGES, TTIntent.ACTION_TIMES_MODIFIED)
        }
        dismiss()
    }

    override fun onDestroyView() {
        binding = null
        if (dialogListener != null) dialogListener!!.onDismissDialog()
        super.onDestroyView()
    }

    companion object {
        fun newInstance(id: Long): TimeDialog {
            val timeDialog = TimeDialog()
            val args = Bundle()
            args.putLong("id", id)
            timeDialog.setArguments(args)
            return timeDialog
        }
    }
}
